plugin/rewrite/name.go

```diff
@@ -16,6 +18,36 @@ func newNameRule(from, to string) (Rule, error) {
 
 // Rewrite rewrites the the current request.
 func (rule *nameRule) Rewrite(w dns.ResponseWriter, r *dns.Msg) Result {
+	requestName := r.Question[0].Name
+	// Rewrite based on substring match
+	if strings.HasPrefix(rule.From, "sub:") {
```

Comment:
This is static to the rule. You should create a constant for matchType (full, substring, regex) and then set it in the rule as a field. Then you don't need string comparisons. Also, you would parse out the `substituteStr` and `regexPattern` and pre-compile the regex at config parse time rather than query processing time.

Reply:
> You should create a constant for matchType (full, substring, regex) and then set it in the rule as a field. 

@johnbelamaric , I like the idea. 

> Also, you would parse out the substituteStr and regexPattern and pre-compile the regex at config parse time rather than query processing time.

Makes sense.